### PR TITLE
PrioGraphScheduler::complete_batch remove TransactionAccountLocks allocation

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
- `TransactionAccountLocks` struct allocates 2 vecs internally, this is not good for performance.
- `unlock_accounts` takes arbitrary iterators over Pubkeys, so we have no need to allocate

#### Summary of Changes
- Iterate over write/read accounts, do not allocate.

Fixes #

### CODE REVIEW

**Comment by apfitzge** [comment]
Unlocking accounts after batch completion is a small portion of overall time, so do not really expect much performance benefit here. Just part of larger effort to reduce allocations.

<details>
<summary>Banking Bench</summary>

```
master:
{'name': 'banking_bench_total', 'median': '51532.09'}
{'name': 'banking_bench_tx_total', 'median': '53217.47'}
{'name': 'banking_bench_success_tx_total', 'median': '50988.14'}

with change:
{'name': 'banking_bench_total', 'median': '51153.70'}
{'name': 'banking_bench_tx_total', 'median': '52860.53'}
{'name': 'banking_bench_success_tx_total', 'median': '50667.78'}
```

</details>

<details>
<summary>bench-tps</summary>
Running with `--num-conflict-groups 1` (with batches of 1, should see more effect of completing batches)

master:
```
[2024-06-17T11:55:28.027346062Z INFO  solana_bench_tps::bench] 
    Average max TPS: 57326.21, 0 nodes had 0 TPS
[2024-06-17T11:55:28.027354608Z INFO  solana_bench_tps::bench] 
    Highest TPS: 57326.21 sampling period 1s max transactions: 3115284 clients: 1 drop rate: 0.62
[2024-06-17T11:55:28.027359457Z INFO  solana_bench_tps::bench] 	Average TPS: 34184.906
```

with change:
```
[2024-06-17T11:50:49.027923929Z INFO  solana_bench_tps::bench] 
    Average max TPS: 59257.83, 0 nodes had 0 TPS
[2024-06-17T11:50:49.027928528Z INFO  solana_bench_tps::bench] 
    Highest TPS: 59257.83 sampling period 1s max transactions: 3239620 clients: 1 drop rate: 0.60
[2024-06-17T11:50:49.027931664Z INFO  solana_bench_tps::bench] 	Average TPS: 35552.043
```

image
Left is with change, right is without the change. 




</details>

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs [hunk 1]
```diff
@@ -358,12 +358,18 @@ impl PrioGraphScheduler {
...
+            let message = transaction.message();
+            let account_keys = message.account_keys();
+            let write_account_locks = account_keys
+                .iter()
+                .enumerate()
+                .filter_map(|(index, key)| message.is_writable(index).then_some(key));
+            let read_account_locks = account_keys
+                .iter()
+                .enumerate()
+                .filter_map(|(index, key)| (!message.is_writable(index)).then_some(key));
```
Comment (ryoqun) [inline core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs:370]: nit: can these useful method be defined on `SanitizedMesssage` or `SanitizedTransaction`?

**Review by ryoqun — APPROVED** [review]
lgtm with nit

File: core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs (hunk 1 above)
Comment (apfitzge) [reply core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs:370]: Thought about it, but I'm generally against adding more stuff into `SDK`. Could probably define a function in this file.

In my experimental branch I do have this defined as member fn of the generic tx trait.

Comment (ryoqun) [reply core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs:370]: okay, thanks for explaining. mine isn't strong opinion.

happy to see this pr merged as-is. :)

**Pull request merged**